subsys:usb:device:audio: Support volume control

SawXu opened the pull request:
Makes USB device audio class support volume control feature.

tmon-nordic commented on subsys/usb/device/class/audio/Kconfig:
```diff
@@ -19,4 +19,25 @@ module = USB_AUDIO
+	default 0
+	range 0 32767
+	help
+	  attributes can range from 32767 down to 0
```
It is somewhat inconsistent that the parameter is of hex type and descriptor uses decimal.

SawXu replied:
> It is somewhat inconsistent that the parameter is of hex type and descriptor uses decimal.

Thanks, I changed to hex type.

tmon-nordic commented on subsys/usb/device/class/audio/audio.c:
```diff
@@ -541,6 +541,73 @@ static int handle_fu_mute_req(struct usb_audio_dev_data *audio_dev_data,
+	uint8_t ch = (setup->wValue) & 0xFF;
+	uint8_t ch_cnt = audio_dev_data->ch_cnt[device];
+	uint8_t *controls = audio_dev_data->controls[device];
+	uint8_t *control_val = &controls[POS(VOLUME, ch, ch_cnt)];
```
Where is VOLUME defined?

SawXu replied:
In `zephyr/subsys/usb/device/class/audio/usb_audio_internal.h` 

You can find these codes.
```

#define POS_MUTE		   0
#define POS_VOLUME		   (POS_MUTE + FEATURE_MUTE_SIZE)
...

#define POS(prop, ch_idx, ch_cnt) (ch_cnt * POS_##prop + \
				  (ch_idx * FEATURE_##prop##_SIZE))
```
It prepends "POS" to prop, so the define is POS_VOLUME

tmon-nordic commented on subsys/usb/device/class/audio/audio.c:
```diff
@@ -541,6 +541,73 @@ static int handle_fu_mute_req(struct usb_audio_dev_data *audio_dev_data,
+			LOG_ERR("*len: %d, LEN(1, VOLUME): %d", *len, LEN(1, VOLUME));
+			return -EINVAL;
+		}
+		if (setup->bRequest == USB_AUDIO_SET_CUR) {
+			evt->val = control_val;
+			evt->val_len = *len;
+			memcpy(control_val, *data, *len);
+			return 0;
+		}
```
The value is not checked against the min and max. Additionally it does not honor the resolution rounding.

Values in USB are little-endian, not target-device-endian. You are missing necessary conversions here (which evalue to no-op on little-endian targets).

SawXu replied:
Thanks a lot.  I fixed it.

tmon-nordic commented on subsys/usb/device/class/audio/Kconfig:
```diff
@@ -19,4 +19,25 @@ module = USB_AUDIO
+	default 0
+	range 0 0x7FFF
+	help
+	  attributes can range from 0x7FFF down to 0
```
The real purpose behind help text is to allow user to get some idea what the value is all about. Talking about some numbers in context of volume is not giving user real idea of what the numbers signify.

USB Device Class Definition for Audio Devices Release 1.0:
> The settings for the CUR, MIN, and MAX attributes can range from +127.9961 dB (0x7FFF) down to -127.9961 dB (0x8001) in steps of 1/256 dB or 0.00390625 dB (0x0001). The range for the CUR attribute is extended by code 0x8000, representing silence, i.e., -∞ dB. The settings for the RES attribute can only take positive values and range from 1/256 dB (0x0001) to +127.9961 dB (0x7FFF). 

I would expect the help text to at least mention that 0x0100 equals to 1 dB.

SawXu replied:
Thanks very much. 
I Updated this file.

jfischer-no commented on subsys/usb/device/class/audio/Kconfig:
```diff
@@ -19,4 +19,31 @@ module = USB_AUDIO
 module-str = USB Audio
 source "subsys/logging/Kconfig.template.log_config"
 
+config USB_DEVICE_AUDIO_MIN_VOLUME
+	hex "USB Audio Device Min Volume"
+	default 0xFA00
+	range 0 0xFFFF
+	help
+	  attention: this attribute is a signed value.
+	  This attribute represents the minimum volume level.
+	  The range from +127.9961 dB (0x7FFF) down to -127.9961 dB (0x8001).
+
+config USB_DEVICE_AUDIO_MAX_VOLUME
+	hex "USB Audio Device Max Volume"
+	default 0x0500
+	range 0 0xFFFF
+	help
+	  attention: this attribute is a signed value.
+	  This attribute represents the maximum volume level.
+	  The range from +127.9961 dB (0x7FFF) down to -127.9961 dB (0x8001).
+
+config USB_DEVICE_AUDIO_RES_VOLUME
+	hex "USB Audio Device Resolution Volume"
+	default 0x100
+	range 0x1 0x7FFF
+	help
+	  attention: this attribute can only take positive values.
+	  The range from 1/256 dB (0x0001) to +127.9961 dB (0x7FFF).
+	  0x100 represents 1dB.
+
```
Audio class are instantiated using devicetree, but the new Kconfig options affect all instances, even if feature-volume does not exist for a particular instance. Is this the right approach?

SawXu replied:
Thanks. How about I move these configs to dts proprites ?

SawXu replied:
I am not sure how to add some properties dependency `hp-feature-volume`.
How about these change in `dts/bindings/usb/usb-audio-hs.yaml`
```
  hp-feature-volume:
    type: boolean
    description: |
      Enable Volume feature.
  volume_max:
    type: int
    default: 1280
    description: |
      todo: add description
  volume_min:
    type: int
    default: -1536
    description: |
      todo: add description
  volume_res:
    type: int
    default: 256
    description: |
      todo: add description
```

jfischer-no replied:
Perhaps better to add new `dts/bindings/usb/usb-audio-feature-volume.yaml` and include it in usb-audio-.*. yaml

SawXu replied:
> Perhaps better to add new `dts/bindings/usb/usb-audio-feature-volume.yaml` and include it in usb-audio-.*. yaml

Thanks, I added these into  `dts/bindings/usb/usb-audio-feature-volume.yaml` and include it in  usb-audio-hs. yaml

But I don't know how to get the corresponding "volume_max" by specific instance in audio.c. And I am not sure which struct these properties should be placed in.

Could you give me some tips please?  Thanks.

jfischer-no replied:
In the drivers we typically have two structures, constant device configuration (dev->config) and runtime device data (dev->data), (unfortunately) audio implementation does not really follow Zephyr Device Driver Model. min, max, and res properties are constant values an likely part of constant device structure, but there is none. You can place min, max, and res in struct usb_audio_dev_data {}, https://github.com/zephyrproject-rtos/zephyr/blob/a18761e29861b709b984eca82dc31efe6214fd62/subsys/usb/device/class/audio/audio.c#L28-L46

or add a pointer to a new constant structure `static const struct feature_volume {uint16_t min; ....};`

Ping me if you are not sure how to get the values from DT properties.

jfischer-no replied:
Thanks, it looks good, I will try to find time for a detailed review in the next few days.

jfischer-no commented on subsys/usb/device/class/audio/audio.c:
```diff
@@ -354,6 +354,19 @@ static void audio_cb_usb_status(struct usb_cfg_data *cfg,
+static int16_t round_vol_to_closest(int16_t target_vol)
+{
+	int rounded_vol  = 0;
+	int16_t step = (int16_t)CONFIG_USB_DEVICE_AUDIO_RES_VOLUME;
```
Why is there (int16_t) cast?

SawXu replied:
This function has been removed since I switched to use ROUND_UP from sys/utils.h

jfischer-no commented on subsys/usb/device/class/audio/audio.c:
```diff
@@ -541,6 +554,81 @@ static int handle_fu_mute_req(struct usb_audio_dev_data *audio_dev_data,
+				return -EINVAL;
+			}
+			if (target_vol % (int16_t)CONFIG_USB_DEVICE_AUDIO_RES_VOLUME != 0) {
+				target_vol = round_vol_to_closest(target_vol);
```
Why cannot ROUND_UP(x, align) form sys/util.h be used here?

jfischer-no commented on subsys/usb/device/class/audio/audio.c:
```diff
@@ -541,6 +554,81 @@ static int handle_fu_mute_req(struct usb_audio_dev_data *audio_dev_data,
+		}
+		if (setup->bRequest == USB_AUDIO_SET_CUR) {
+			target_vol = *((int16_t *)*data);
+			if (target_vol > (int16_t)CONFIG_USB_DEVICE_AUDIO_MAX_VOLUME ||
```
Why (int16_t)?

Can IN_RANGE be used here?
```c

		!IN_RANGE(target_vol,
			  CONFIG_USB_DEVICE_AUDIO_MIN_VOLUME,
			  CONFIG_USB_DEVICE_AUDIO_MAX_VOLUME)
```

SawXu replied:
changed to use IN_RANGE

tmon-nordic commented on subsys/usb/device/class/audio/audio.c:
```diff
@@ -541,6 +556,77 @@ static int handle_fu_mute_req(struct usb_audio_dev_data *audio_dev_data,
+			return 0;
+		}
+	} else {
+		if (setup->bRequest == USB_AUDIO_GET_CUR) {
+			*len = LEN(ch_cnt, VOLUME);
+			memcpy(*data, control_val, *len);
+			return 0;
+		} else if (setup->bRequest == USB_AUDIO_GET_MIN) {
+			*len = sizeof(audio_dev_data->volumes.volume_min);
+			memcpy(*data, &(audio_dev_data->volumes.volume_min), *len);
+			return 0;
+		} else if (setup->bRequest == USB_AUDIO_GET_MAX) {
+			*len = sizeof(audio_dev_data->volumes.volume_max);
+			memcpy(*data, &(audio_dev_data->volumes.volume_max), *len);
+			return 0;
+		} else if (setup->bRequest == USB_AUDIO_GET_RES) {
+			*len = sizeof(audio_dev_data->volumes.volume_res);
+			memcpy(*data, &(audio_dev_data->volumes.volume_res), *len);
+			return 0;
+		}
```
The memcpy here does not take the fact that USB uses Little-Endian into account. While it will work on LE targets, the code will give incorrect results on BE targets.

SawXu replied:
Thanks, I added Endian convert.

SawXu committed:
dts: bindings: usb_audio: Add volume properties

Add volume feature specific fields. So remove this outdated description.

Signed-off-by: Saw Xu <Saw1993@126.com>

SawXu committed:
samples: subsys: usb: audio: Add volume control case

Add sample code for getting volume to headset sample and
headphone_microphone sample

Signed-off-by: Saw Xu <Saw1993@126.com>

SawXu committed:
usb: device: audio: Support volume control

Makes USB device audio class support volume control feature.

Signed-off-by: Saw Xu <Saw1993@126.com>

jfischer-no commented on subsys/usb/device/class/audio/audio.c:
```diff
@@ -541,6 +556,82 @@ static int handle_fu_mute_req(struct usb_audio_dev_data *audio_dev_data,
 	return -EINVAL;
 }
 
+/**
+ * @brief Handler for feature unit volume control requests.
+ *
+ * This function handles feature unit volume control request.
+ *
+ * @param audio_dev_data USB audio device data.
+ * @param setup          Information about the executed request.
+ * @param len		 Size of the buffer.
+ * @param data		 Buffer containing the request result.
+ * @param evt		 Feature Unit Event info.
+ * @param device	 Device part that has been addressed. Applicable for
+ *			 bidirectional device.
+ *
+ * @return 0 if successful, negative errno otherwise.
+ */
```
Sorry, I completely lost track. You can drop the doxygen comments, they are not used here anyway.  The others should also be removed but it is my fault that they are there at all.

SawXu replied:
OK, I have removed the comments.
